src/expand.ts

```diff
@@ -0,0 +1,49 @@
+import { Filter, OrExp } from "./index";
+import { valfilter } from "./flatten";
+
+const isOr = (f: Filter): f is OrExp => f.op === 'or';
+
+function baseProduct<T>(previousProduct: T[][], b: T[]): T[][] {
+  const result = [];
+
+  for (const x of previousProduct) {
+    for (const y of b) {
+      result.push([...x, y]);
+    }
+  }
+
+  return result;
+}
+
+function product<T>(arrays: T[][]): T[][] {
+  const [first, ...rest] = arrays;
+
+  let result = first.map((v) => [v]);
+
+  for (const array of rest) {
+    result = baseProduct(result, array);
+  }
+
+  return result;
+}
```

Comment:
We could probably be more efficient and avoid creating intermediate arrays here, but I found this formulation the most readable way to define the cartesian product of arrays.